src/pycapsule.rs

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+use crate::Python;
+use crate::{ffi, AsPyPointer, PyAny};
+use crate::{pyobject_native_type_core, PyErr, PyResult};
+use std::ffi::{c_void, CStr};
+use std::os::raw::c_int;
+
+/// Represents a Python Capsule
+/// As described in [Capsules](https://docs.python.org/3/c-api/capsule.html#capsules)
+/// > This subtype of PyObject represents an opaque value, useful for C extension
+/// > modules who need to pass an opaque value (as a void* pointer) through Python
+/// > code to other C code. It is often used to make a C function pointer defined
+/// > in one module available to other modules, so the regular import mechanism can
+/// > be used to access C APIs defined in dynamically loaded modules.
+///
+///
+/// # Example
+/// ```
+///  use std::ffi::CString;
+///  use pyo3::{prelude::*, pycapsule::PyCapsule};
+///
+///  #[repr(C)]
+///  struct Foo {
+///     pub val: u32,
+///  }
+///
+///  let r = Python::with_gil(|py| -> PyResult<()> {
+///     let foo = Foo { val: 123 };
+///     let name = CString::new("builtins.capsule").unwrap();
+///
+///     let capsule = PyCapsule::new(py, foo, name.as_ref(), None)?;
+///
+///     let module = PyModule::import(py, "builtins")?;
+///     module.add("capsule", capsule)?;
+///
+///     let cap: &Foo = PyCapsule::import(py, name.as_ref(), false)?;
+///     assert_eq!(cap.val, 123);
+///     Ok(())
+///  });
+///  assert!(r.is_ok());
+/// ```
+#[repr(transparent)]
+pub struct PyCapsule(PyAny);
+
+pyobject_native_type_core!(PyCapsule, ffi::PyCapsule_Type, #checkfunction=ffi::PyCapsule_CheckExact);
+
+impl PyCapsule {
+    /// Constructs a new capsule of whose contents are `T` associated with `name`.
+    /// Can optionally provide a deconstructor for when `PyCapsule` is destroyed
+    /// it will be passed the capsule.
+    pub fn new<'py, T>(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+        destructor: Option<ffi::PyCapsule_Destructor>,
```

Comment:
I think that asking the user to provide the destructor which receives `*mut void` is hard for users to use correctly; for example the user should use `Box::from_raw` to avoid leaking the value, but that's not really obvious at all. Also possible panics and maybe GIL interactions.

One possible alternative would be to ask users to pass a destructor which takes T:

```rust
destructor: Option<impl FnOnce(T)>
```

And then we can box this up with the value (instead of just boxing the value), and internally create the "real" destructor which first unboxes the value and calls the closure:

```rust
struct ValueAndDestructor<T, D: FnOnce(T)> {
    value: T,
    destructor: D
}

unsafe extern "C" fn capsule_destructor<T, F: FnOnce<T>>(capsule: *mut PyObject) {
    // FIXME might need to use `handle_panic` here and create a gilpool?
    let ptr = unsafe { PyCapsule_GetPointer(capsule, PyCapsule_GetName(capsule)) };
    let ValueAndDestructor { value, destructor } = *unsafe { Box::from_raw(ptr as *mut ValueAndDestructor<T, F>) };
    destructor(value)
}
```

And then I think if the user does not provide a destructor, would default to `std::mem::drop` or some other no-op method.

... This design doesn't take into account the `context` pointer which can also be stored in the capsule (or it's name). Pehaps the capsule could be a second argument to the `FnOnce`?

Reply:
I've updated it with a similar implementation, but I end up being confused. 

The pointer value of the capsule is pointing to a struct for `T` and `F`, and _somehow_ `get_reference<T>(&self) -> &T` still works, even though it's really just getting the pointer to this `CapsuleContents` struct that holds the value _and_ destructor. Maybe it's some Rust/memory internals I don't know about yet that allows that to work.

I was thinking it ought to be something more like:
```rust
pub unsafe fn reference<T, F>(&self) -> &T {
   let contents = Box::from_raw(self.get_pointer() as *mut &CapsuleContents<T, F> };
   &contents.value
}
```

but again, it works as-is. And even if it didn't, would this be good? Seems having to supply type parameters for everything in `CapsuleContents` would be annoying, especially for methods like `reference` or `get_destructor` where the user is only interested in getting a single value, and yet required to supply types for all values in a capsule. Or maybe it isn't that bad...

Do you think having a macro of some sort for a destructor would be helpful for the user?
ie:

```rust
#[capsule_destructor]
fn destructor(capsule: PyCapsule) {
    ...smart stuff...
}

// expands to something like
unsafe extern "C" fn capsule_destructor(capsule: *mut PyObject) {
    let ptr = unsafe { PyCapsule_GetPointer(capsule, PyCapsule_GetName(capsule)) };
    let capsule = *Box::from_raw(ptr as *mut PyCapsule);
    ...smart stuff...
}
```